[WebXR][OpenXR] Add EGL fences for WebGL content

svillar opened the pull request:
#### b50c48fd07c7c53424ac797ef8bafc11a81f2b7c
<pre>
[WebXR][OpenXR] Add EGL fences for WebGL content
<a href="https://bugs.webkit.org/show_bug.cgi?id=297140">https://bugs.webkit.org/show_bug.cgi?id=297140</a>

Reviewed by Nikolas Zimmermann.

Currently there is no guarantee that the WebXR WebGL frames are
completely rendered before they&apos;re submitted to the OpenXR compositor.
That&apos;s why the WebXROpaqueFramebuffer does a glFinish(), blocking the
WebProcess before returning from endFrame.

A much better solution is to use EGL fences. The WebProcess will
create a fence in endFrame(). Then that fence is exported and shared
to the UI process (via a file descriptor) in the submitFrame() call.

The UI process then will call wait in the fence before submitting
the frame to OpenXR. Although OpenXR runtimes guarantee that all the
GPU commands are finished before displaying the contents of the
textures, that does not apply to the cases in which there are multiple
contexts accessing the texture, as it happens in WebKit due to the
multiprocess architecture.

Co-authored-by: Carlos Garcia Campos &lt;cgarcia@igalia.com&gt;
Canonical link: <a href="https://commits.webkit.org/299154@main">https://commits.webkit.org/299154@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46b6c50b5dca20e9b835bafc57ae145b55d428da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118048 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37726 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28362 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124198 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70084 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38419 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46308 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89582 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59183 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121000 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30603 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105852 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70074 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23969 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67863 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100026 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24147 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127279 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44951 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33874 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98254 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45312 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102076 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98040 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43442 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21431 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41387 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18813 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44823 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50497 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44283 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47628 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45972 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -203,7 +204,7 @@ void WebXROpaqueFramebuffer::startFrame(PlatformXR::FrameData::LayerData& data)
 }
 
-void WebXROpaqueFramebuffer::endFrame()
+void WebXROpaqueFramebuffer::endFrame(std::optional<WTF::UnixFileDescriptor>& fenceFD)
```
No need to use optional for a fd, it will be -1 when not used.

carlosgcampos commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -225,7 +226,13 @@ void WebXROpaqueFramebuffer::endFrame()
+#if HAVE(GL_FENCE)
+    if (auto fence =  GLFenceEGL::createExportable()) {
+        gl->flush();
+        fenceFD = fence->exportFD();
```
You shouldn't need to do the flush, GLFence already calls glFlush() right after the fence is created. Also in case exportable is not supported, you can still create a non-exportable fence and call clientWait to avoid the glFinish. Thisd is what we do in AcceleratedSurface.cpp and RemoteGraphicsContextGLGBM.cpp

kkinnunen-apple replied:
As explained by Dan, all the general purpose code, such as in Modules/webxr, should refer to the abstract ExternalSync object.
Your platform can then implement them in any way you want. The platform can even have multiple different types of sync objects.

The general purpose logic should be similar for all the platforms. In case the general purpose code is distinct to each platform, it mostly sounds like a code smell. In general, it does not make sense that WPE/GTK ports would need a particular synchronisation when Cocoa would not. So care needs to be taken when adding ifdefing.

svillar replied:
@kkinnunen-apple so the difference with cocoa port is that in Cocoa, the UI process can create a match metal object that will be automatically signaled when the fence is signaled because you use a Metal specific extension for that (EGL_SYNC_METAL_SHARED_EVENT_OBJECT_ANGLE). 

In our case, as there is no such a thing, we just need to export the fence via a file descriptor to let the UI process wait for the signaling.

svillar replied:
Anyway I'll modify this to use GCGLExternalSync. Thanks for the review

carlosgcampos replied:
I've updated the patch implementing GLExternalSync

carlosgcampos commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
```diff
@@ -102,7 +102,11 @@ class WebXROpaqueFramebuffer {
 
     void startFrame(PlatformXR::FrameData::LayerData&);
+#if PLATFORM(WPE) || PLATFORM(GTK)
+    void endFrame(std::optional<WTF::UnixFileDescriptor>&);
```
I think it's clearer if the fd is the return value instead of an output parameter

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -277,6 +277,54 @@ void OpenXRCoordinator::scheduleAnimationFrame(WebPageProxy& page, std::optional
+    auto isEGL15 = display.eglCheckVersion(1, 5);
+    EGLSync sync { nullptr };
+    Vector<EGLAttrib> attributes = {
+        EGL_SYNC_NATIVE_FENCE_FD_ANDROID, WTFMove(fenceFD).release(),
```
```suggestion
        EGL_SYNC_NATIVE_FENCE_FD_ANDROID, fenceFD.release(),
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -277,6 +277,54 @@ void OpenXRCoordinator::scheduleAnimationFrame(WebPageProxy& page, std::optional
+        if (isEGL15)
+            eglClientWaitSync(display.eglDisplay(), sync, 0, EGL_FOREVER);
+        else
+            eglClientWaitSyncKHR(display.eglDisplay(), sync, 0, EGL_FOREVER_KHR);
```
I'm not sure this is a good idea, if server wait is not supported, it's better to wait in the web process or asynchronously poll the fd, because this client wait will block the UI process. For now we can probably assume that if exportable is supported, server wait is too.

djg reviewed (changes requested):
Per my comment.

svillar commented on Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
```diff
@@ -304,14 +304,33 @@ void WebXRWebGLLayer::startFrame(PlatformXR::FrameData& data)
+            fenceFD = gl->exportExternalSync(sync);
+            gl->deleteExternalSync(sync);
+        }
+    }
```
I think the idea is to encapsulate the external sync management in the opaque frame buffer class. Why do you need to move it here?

carlosgcampos replied:
It's not moved, the opaque frame buffer is the one creating the the external sync, but here is where we need to export the sync into an fd that we can pass to the UI process.

carlosgcampos replied:
Ok, managed to find a simpler solution in the end, by storing the fence fd in the opaque buffer.

svillar commented on Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
```diff
@@ -95,6 +95,14 @@ GraphicsContextGLANGLE::~GraphicsContextGLANGLE()
+        for (auto* sync : m_eglSyncs.values()) {
+            bool result = EGL_DestroySync(m_displayObj, sync);
+            ASSERT_UNUSED(result, !!result);
+        }
```
Wouldn't the presence of images/syncs it be more an error condition than something we need to handle here?

carlosgcampos replied:
We are already releasing all other gl resources here. Other ports do the same.

svillar commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -225,7 +225,16 @@ void WebXROpaqueFramebuffer::endFrame()
+        gl->deleteExternalSync(sync);
+    } else
+        gl->finish();
+}
```
That's nicer indeed, could we add an ASSERT(!m_fenceFD) in `startFrame` just to be sure that we are not missing any potential issue where the fd is not taken just after endFrame (which should happen as the WebXRSession calls endFrame just before submitFrame).

nikolaszimmermann reviewed (approved):
Looks really nice. r=me.

svillar committed:
[WebXR][OpenXR] Add EGL fences for WebGL content
https://bugs.webkit.org/show_bug.cgi?id=297140

Reviewed by Nikolas Zimmermann.

Currently there is no guarantee that the WebXR WebGL frames are
completely rendered before they're submitted to the OpenXR compositor.
That's why the WebXROpaqueFramebuffer does a glFinish(), blocking the
WebProcess before returning from endFrame.

A much better solution is to use EGL fences. The WebProcess will
create a fence in endFrame(). Then that fence is exported and shared
to the UI process (via a file descriptor) in the submitFrame() call.

The UI process then will call wait in the fence before submitting
the frame to OpenXR. Although OpenXR runtimes guarantee that all the
GPU commands are finished before displaying the contents of the
textures, that does not apply to the cases in which there are multiple
contexts accessing the texture, as it happens in WebKit due to the
multiprocess architecture.

Co-authored-by: Carlos Garcia Campos <cgarcia@igalia.com>
Canonical link: https://commits.webkit.org/299154@main
